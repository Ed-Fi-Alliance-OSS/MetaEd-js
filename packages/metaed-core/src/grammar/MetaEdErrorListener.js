// @flow
import antlr4 from 'antlr4';
import type { ValidationFailure } from '../validator/ValidationFailure';

export class MetaEdErrorListener {
  messageCollection: ValidationFailure[];
  validatorName: string;

  constructor(messageCollection: ValidationFailure[], validatorName: string = 'MetaEdErrorListener') {
    antlr4.error.ErrorListener.call(this);
    this.messageCollection = messageCollection;
    this.validatorName = validatorName;
  }

  syntaxError(
    recognizer: any,
    offendingSymbol: any,
    concatenatedLineNumber: number,
    characterPosition: number,
    message: string,
  ) {
    this.messageCollection.push({
      validatorName: this.validatorName,
      category: 'error',
      message,
      sourceMap: {
        line: concatenatedLineNumber,
        column: characterPosition,
        tokenText: offendingSymbol && offendingSymbol.text ? offendingSymbol.text : '',
      },
      fileMap: null,
    });
  }

  getMessageCollection(): ValidationFailure[] {
    return this.messageCollection;
  }
}
