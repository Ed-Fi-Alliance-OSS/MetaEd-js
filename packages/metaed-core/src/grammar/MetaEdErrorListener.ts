// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import antlr4 from '@edfi/antlr4';
import { ValidationFailure } from '../validator/ValidationFailure';

export class MetaEdErrorListener {
  messageCollection: ValidationFailure[];

  validatorName: string;

  constructor(messageCollection: ValidationFailure[], validatorName: string = 'MetaEdErrorListener') {
    antlr4.error.ErrorListener.call(this);
    this.messageCollection = messageCollection;
    this.validatorName = validatorName;
  }

  syntaxError(_1: any, offendingSymbol: any, concatenatedLineNumber: number, characterPosition: number, message: string) {
    this.messageCollection.push({
      validatorName: this.validatorName,
      category: 'error',
      message,
      sourceMap: {
        line: concatenatedLineNumber,
        column: characterPosition,
        tokenText: offendingSymbol && offendingSymbol.text ? offendingSymbol.text : '',
      },
      fileMap: null,
    });
  }

  getMessageCollection(): ValidationFailure[] {
    return this.messageCollection;
  }
}
