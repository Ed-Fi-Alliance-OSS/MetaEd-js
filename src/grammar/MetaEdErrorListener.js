// @flow
import antlr4 from 'antlr4';
import type { ValidationMessage } from '../core/validators/ValidationTypes';
import type { FileIndex, FilenameAndLineNumber } from '../core/tasks/FileIndex';
import { getFilenameAndLineNumber } from '../core/tasks/FileIndex';

export default class MetaEdErrorListener {
  _messageCollection: ValidationMessage[];
  _fileIndex: FileIndex;

  constructor(messageCollection: ValidationMessage[], fileIndex: FileIndex) {
    antlr4.error.ErrorListener.call(this);
    this._messageCollection = messageCollection;
    this._fileIndex = fileIndex;
  }

  syntaxError(recognizer: any, offendingSymbol: any, concatenatedLineNumber: number, characterPosition: number,
    message: string /* , e */) {
    const metaEdFile: FilenameAndLineNumber = getFilenameAndLineNumber(this._fileIndex, concatenatedLineNumber);

    this._messageCollection.push({
      message,
      characterPosition,
      concatenatedLineNumber,
      filename: metaEdFile.filename,
      lineNumber: metaEdFile.lineNumber,
      tokenText: offendingSymbol && offendingSymbol.text ? offendingSymbol.text : '',
    });
  }

  getMessageCollection(): ValidationMessage[] {
    return this._messageCollection;
  }
}
