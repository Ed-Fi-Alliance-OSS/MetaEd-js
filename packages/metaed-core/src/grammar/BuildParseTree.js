// @flow
import R from 'ramda';
import winston from 'winston';
import { MetaEdErrorListener } from './MetaEdErrorListener';
import type { State } from '../State';
import { getAllContents, getFilenameAndLineNumber } from '../file/FileIndex';
import type { ParseTreeBuilder } from './ParseTreeBuilder';

export const buildParseTree = R.curry(
  (parseTreeBuilder: ParseTreeBuilder, state: State): void => {
    const validationFailures = [];

    const errorListener = new MetaEdErrorListener(validationFailures, 'BuildParseTree - MetaEdErrorListener');
    const parseTree = parseTreeBuilder(errorListener, getAllContents(state.fileIndex));

    if (parseTree == null) {
      winston.error('BuildParseTree: parse tree builder returned null for state metaEdFileIndex contents');
    }

    validationFailures.forEach(failure => {
      if (failure.sourceMap && state.fileIndex) {
        failure.fileMap = getFilenameAndLineNumber(state.fileIndex, failure.sourceMap.line);
      }
    });

    //    state.validationFailure.push(...validationFailures);

    // eslint-disable-next-line no-param-reassign
    state.parseTree = parseTree;
  },
);
