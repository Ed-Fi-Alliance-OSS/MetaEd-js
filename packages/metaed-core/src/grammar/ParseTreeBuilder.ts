import antlr4 from '@edfi/antlr4/index';
import { BaseLexer } from './gen/BaseLexer';
import { MetaEdGrammar } from './gen/MetaEdGrammar';
import { MetaEdErrorListener } from './MetaEdErrorListener';

export type ParseTreeBuilder = (metaEdErrorListener: MetaEdErrorListener, metaEdContents: string) => MetaEdGrammar;

function errorListeningParser(metaEdErrorListener: MetaEdErrorListener, metaEdContents: string): MetaEdGrammar {
  const lexer = new BaseLexer(new antlr4.InputStream(metaEdContents));
  const parser = new MetaEdGrammar(new antlr4.CommonTokenStream(lexer));
  lexer.removeErrorListeners();
  lexer.addErrorListener(metaEdErrorListener);
  parser.removeErrorListeners();
  parser.addErrorListener(metaEdErrorListener);
  return parser;
}

export function buildMetaEd(metaEdErrorListener: MetaEdErrorListener, metaEdContents: string): MetaEdGrammar {
  return errorListeningParser(metaEdErrorListener, metaEdContents).metaEd();
}

export function buildTopLevelEntity(metaEdErrorListener: MetaEdErrorListener, metaEdContents: string): MetaEdGrammar {
  return errorListeningParser(metaEdErrorListener, metaEdContents).topLevelEntity();
}
