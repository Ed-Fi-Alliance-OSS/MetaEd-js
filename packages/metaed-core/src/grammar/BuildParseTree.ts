import * as R from 'ramda';
import winston from 'winston';
import type MetaEdGrammar from './gen/MetaEdGrammar';
import { ValidationFailure } from '../validator/ValidationFailure';
import { MetaEdErrorListener } from './MetaEdErrorListener';
import { State } from '../State';
import { getAllContents, getFilenameAndLineNumber } from '../file/FileIndex';
import { ParseTreeBuilder } from './ParseTreeBuilder';

export const buildParseTree = R.curry((parseTreeBuilder: ParseTreeBuilder, state: State): void => {
  const validationFailures: ValidationFailure[] = [];

  const errorListener = new MetaEdErrorListener(validationFailures, 'BuildParseTree - MetaEdErrorListener');
  const parseTree: MetaEdGrammar = parseTreeBuilder(errorListener, getAllContents(state.fileIndex));

  if (parseTree == null) {
    winston.error('BuildParseTree: parse tree builder returned null for state metaEdFileIndex contents');
  }

  validationFailures.forEach((failure) => {
    if (failure.sourceMap && state.fileIndex) {
      failure.fileMap = getFilenameAndLineNumber(state.fileIndex, failure.sourceMap.line);
    }
  });

  state.parseTree = parseTree.metaEd();
});
