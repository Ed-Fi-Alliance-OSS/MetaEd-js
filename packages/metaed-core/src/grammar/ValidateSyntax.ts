// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import * as R from 'ramda';
import { State } from '../State';
import { ValidationFailure } from '../validator/ValidationFailure';
import { MetaEdFile, FileSet } from '../file/MetaEdFile';
import { createFileIndex, getFilenameAndLineNumber } from '../file/FileIndex';
import { MetaEdErrorListener } from './MetaEdErrorListener';
import { ParseTreeBuilder } from './ParseTreeBuilder';
import { Logger } from '../Logger';

export const validateSyntax = R.curry((parseTreeBuilder: ParseTreeBuilder, state: State): void => {
  if (state.loadedFileSet == null) {
    Logger.error('ValidateSyntax: no files to load found');
    return;
  }

  state.loadedFileSet.forEach((fileToLoad: FileSet) => {
    fileToLoad.files.forEach((file: MetaEdFile) => {
      const validationFailures: ValidationFailure[] = [];
      const errorListener = new MetaEdErrorListener(validationFailures, 'ValidateSyntax - MetaEdErrorListener');

      const parseTree = parseTreeBuilder(errorListener, file.contents);
      if (parseTree == null) {
        Logger.error(`ValidateSyntax: parse tree builder returned null for file ${file.fullPath}`);
      }

      validationFailures.forEach((failure: ValidationFailure) => {
        if (failure.sourceMap != null) {
          const lineNumber: number = failure.sourceMap.line;
          failure.fileMap = getFilenameAndLineNumber(createFileIndex([file]), lineNumber);
        }
      });

      state.validationFailure.push(...validationFailures);
    });
  });
});
