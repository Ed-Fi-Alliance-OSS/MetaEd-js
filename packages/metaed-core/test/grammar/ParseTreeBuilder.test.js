// @flow
import { MetaEdErrorListener } from '../../src/grammar/MetaEdErrorListener';
import { buildTopLevelEntity, buildMetaEd } from '../../src/grammar/ParseTreeBuilder';
import type { ValidationFailure } from '../../src/validator/ValidationFailure';

describe('when parsing top level entities', () => {
  it('should parse with valid MetaEd', () => {
    const inputText = [
      'Domain Entity TestEntity',
      'documentation "This is the first line\nThis is more..."',
      '    integer MyProperty',
      '        documentation "Integer documentation"',
      '        is part of identity\n',
    ].join('\n');

    const validationFailures: Array<ValidationFailure> = [];
    const errorListener = new MetaEdErrorListener(validationFailures);
    buildTopLevelEntity(errorListener, inputText);
    expect(validationFailures).toHaveLength(0);
  });

  it('should not parse with extraneous xyz', () => {
    const inputText = [
      'Domain Entity TestEntity',
      'documentation "This is the first line\nThis is more..."',
      '    integer MyProperty xyz',
      '        documentation "Integer documentation"',
      '        is part of identity\n',
    ].join('\n');

    const validationFailures: Array<ValidationFailure> = [];
    const errorListener = new MetaEdErrorListener(validationFailures);
    buildTopLevelEntity(errorListener, inputText);
    expect(validationFailures).toHaveLength(1);
    expect(validationFailures[0].message).toMatchSnapshot('when parsing top level entities should not parse with extraneous xyz - message');
    expect(validationFailures[0].sourceMap).toMatchSnapshot('when parsing top level entities should not parse with extraneous xyz - sourceMap');
  });

  it('should not parse with multiple keyword syntax errors', () => {
    const inputText = [
      'Domain Entity Staff additions\n',
      'descriptor EducatorEffectiveness\n',
      'documentation "Indicates the educator effectiveness label as identified by the school district for each educator. The effectiveness labels should be based upon the employees most recent evaluation."\n',
      'is optional\n',
      'comvmon Recognition\n',
      'documentation "Recognition given to a staff member for accomplishments in a co-curricular, or extra-curricular activity."\n',
      'is optional collection\n',
      'shared decimal ProfessionalDevelopment named NewTeacherClassroomManagementProfessionalDevelopment\n',
      'documentation "The number of days of professional development in classroom management that was provided by the district for new teacher."\n',
      'is optional\n',
      'shared decimal ProfessionalDevelopment named NewTeacherInstructionalDeliveryProfessionalDevelopment\n',
      'documentation "The number of days of professional development in instructional deliver (strategies) that was provided by the district for new teacher."\n',
      'is optional\n',
      'zcommvzxxon REPRaceOrder\n',
      'documentation "none provided"\n',
      'is optional collection\n',
    ].join('\n');

    const validationFailures: Array<ValidationFailure> = [];
    const errorListener = new MetaEdErrorListener(validationFailures);
    buildTopLevelEntity(errorListener, inputText);
    expect(validationFailures).toHaveLength(2);
    expect(validationFailures[0].message).toMatchSnapshot('when parsing top level entities should not parse with multiple keyword syntax errors - message');
    expect(validationFailures[0].sourceMap).toMatchSnapshot('when parsing top level entities should not parse with multiple keyword syntax errors - sourceMap');
    expect(validationFailures[1].message).toMatchSnapshot('when parsing top level entities should not parse with multiple keyword syntax errors - message');
    expect(validationFailures[1].sourceMap).toMatchSnapshot('when parsing top level entities should not parse with multiple keyword syntax errors - sourceMap');
  });

  it('another example', () => {
    const inputText = [
      'Domain Entity Staff additions\n',
      'descriptor EducatorEffectiveness\n',
      'documentation "Indicates the educator effectiveness label as identified by the school district for each educator. The effectiveness labels should be based upon the employees most recent evaluation."\n',
      'is optional\n',
      'comvmon Recognition\n',
      'documentation "Recognition given to a staff member for accomplishments in a co-curricular, or extra-curricular activity."\n',
      'is optional collection\n',
    ].join('\n');

    const validationFailures: Array<ValidationFailure> = [];
    const errorListener = new MetaEdErrorListener(validationFailures);
    buildTopLevelEntity(errorListener, inputText);
    expect(validationFailures).toHaveLength(1);
    expect(validationFailures[0].message).toMatchSnapshot('when parsing top level entities another example - message');
    expect(validationFailures[0].sourceMap).toMatchSnapshot('when parsing top level entities another example - sourceMap');
  });

  it('and another example', () => {
    const inputText = [
      'Domain Entity Staff\n',
      'documentation "Something"\n',
      'descriptor EducatorEffectiveness\n',
      'documentation "Indicates the educator effectiveness label as identified by the school district for each educator. The effectiveness labels should be based upon the employees most recent evaluation."\n',
      'is optional\n',
      'comvmon Recognition\n',
      'documentation "Recognition given to a staff member for accomplishments in a co-curricular, or extra-curricular activity."\n',
      'is optional collection\n',
    ].join('\n');

    const validationFailures: Array<ValidationFailure> = [];
    const errorListener = new MetaEdErrorListener(validationFailures);
    buildTopLevelEntity(errorListener, inputText);
    expect(validationFailures).toHaveLength(1);
    expect(validationFailures[0].message).toMatchSnapshot('when parsing top level entities and another example - message');
    expect(validationFailures[0].sourceMap).toMatchSnapshot('when parsing top level entities and another example - sourceMap');
  });

  it('one more example', () => {
    const inputText = [
      'Domain Entity Staff\n',
      'documentation "Something"\n',
      'decscriptor EducatorEffectiveness\n',
      'documentation "Indicates the educator effectiveness label as identified by the school district for each educator. The effectiveness labels should be based upon the employees most recent evaluation."\n',
      'is optional\n',
      'comvmon Recognition\n',
      'documentation "Recognition given to a staff member for accomplishments in a co-curricular, or extra-curricular activity."\n',
      'is optional collection\n',
    ].join('\n');

    const validationFailures: Array<ValidationFailure> = [];
    const errorListener = new MetaEdErrorListener(validationFailures);
    buildTopLevelEntity(errorListener, inputText);
    expect(validationFailures).toHaveLength(1);
    expect(validationFailures[0].message).toMatchSnapshot('when parsing top level entities one more example - message');
    expect(validationFailures[0].sourceMap).toMatchSnapshot('when parsing top level entities one more example - sourceMap');
  });
});

describe('when parsing full MetaEd', () => {
  it('should parse with valid MetaEd', () => {
    const inputText = [
      'Begin Namespace extension EXTENSION',
      'Domain Entity TestEntity',
      'documentation "This is the first line\nThis is more..."',
      '    integer MyProperty',
      '        documentation "Integer documentation"',
      '        is part of identity\n',
      'End Namespace',
    ].join('\n');

    const validationFailures: Array<ValidationFailure> = [];
    const errorListener = new MetaEdErrorListener(validationFailures);
    buildMetaEd(errorListener, inputText);
    expect(validationFailures).toHaveLength(0);
  });

  it('should not parse with extraneous xyz', () => {
    const inputText = [
      'Begin Namespace extension EXTENSION',
      'Domain Entity TestEntity',
      'documentation "This is the first line\nThis is more..."',
      '    integer MyProperty xyz',
      '        documentation "Integer documentation"',
      '        is part of identity\n',
      'End Namespace',
    ].join('\n');

    const validationFailures: Array<ValidationFailure> = [];
    const errorListener = new MetaEdErrorListener(validationFailures);
    buildMetaEd(errorListener, inputText);
    expect(validationFailures).toHaveLength(1);
    expect(validationFailures[0].message).toMatchSnapshot('when parsing full MetaEd should not parse with extraneous xyz - message');
    expect(validationFailures[0].sourceMap).toMatchSnapshot('when parsing full MetaEd should not parse with extraneous xyz - sourceMap');
  });

  it('should not parse with multiple keyword syntax errors', () => {
    const inputText = [
      'Begin Namespace extension EXTENSION',
      'Domain Entity Staff additions\n',
      'descriptor EducatorEffectiveness\n',
      'documentation "Indicates the educator effectiveness label as identified by the school district for each educator. The effectiveness labels should be based upon the employees most recent evaluation."\n',
      'is optional\n',
      'comvmon Recognition\n',
      'documentation "Recognition given to a staff member for accomplishments in a co-curricular, or extra-curricular activity."\n',
      'is optional collection\n',
      'shared decimal ProfessionalDevelopment named NewTeacherClassroomManagementProfessionalDevelopment\n',
      'documentation "The number of days of professional development in classroom management that was provided by the district for new teacher."\n',
      'is optional\n',
      'shared decimal ProfessionalDevelopment named NewTeacherInstructionalDeliveryProfessionalDevelopment\n',
      'documentation "The number of days of professional development in instructional deliver (strategies) that was provided by the district for new teacher."\n',
      'is optional\n',
      'zcommvzxxon REPRaceOrder\n',
      'documentation "none provided"\n',
      'is optional collection\n',
      'End Namespace',
    ].join('\n');

    const validationFailures: Array<ValidationFailure> = [];
    const errorListener = new MetaEdErrorListener(validationFailures);
    buildMetaEd(errorListener, inputText);
    expect(validationFailures).toHaveLength(2);
    expect(validationFailures[0].message).toMatchSnapshot('when parsing full MetaEd should not parse with multiple keyword syntax errors - message');
    expect(validationFailures[0].sourceMap).toMatchSnapshot('when parsing full MetaEd should not parse with multiple keyword syntax errors - sourceMap');
    expect(validationFailures[1].message).toMatchSnapshot('when parsing full MetaEd should not parse with multiple keyword syntax errors - message');
    expect(validationFailures[1].sourceMap).toMatchSnapshot('when parsing full MetaEd should not parse with multiple keyword syntax errors - sourceMap');
  });

  it('another example', () => {
    const inputText = [
      'Begin Namespace extension EXTENSION',
      'Domain Entity Staff additions\n',
      'descriptor EducatorEffectiveness\n',
      'documentation "Indicates the educator effectiveness label as identified by the school district for each educator. The effectiveness labels should be based upon the employees most recent evaluation."\n',
      'is optional\n',
      'comvmon Recognition\n',
      'documentation "Recognition given to a staff member for accomplishments in a co-curricular, or extra-curricular activity."\n',
      'is optional collection\n',
      'End Namespace',
    ].join('\n');

    const validationFailures: Array<ValidationFailure> = [];
    const errorListener = new MetaEdErrorListener(validationFailures);
    buildMetaEd(errorListener, inputText);
    expect(validationFailures).toHaveLength(1);
    expect(validationFailures[0].message).toMatchSnapshot('when parsing full MetaEd another example - message');
    expect(validationFailures[0].sourceMap).toMatchSnapshot('when parsing full MetaEd another example - sourceMap');
  });

  it('and another example', () => {
    const inputText = [
      'Begin Namespace extension EXTENSION',
      'Domain Entity Staff\n',
      'documentation "Something"\n',
      'descriptor EducatorEffectiveness\n',
      'documentation "Indicates the educator effectiveness label as identified by the school district for each educator. The effectiveness labels should be based upon the employees most recent evaluation."\n',
      'is optional\n',
      'comvmon Recognition\n',
      'documentation "Recognition given to a staff member for accomplishments in a co-curricular, or extra-curricular activity."\n',
      'is optional collection\n',
      'End Namespace',
    ].join('\n');

    const validationFailures: Array<ValidationFailure> = [];
    const errorListener = new MetaEdErrorListener(validationFailures);
    buildMetaEd(errorListener, inputText);
    expect(validationFailures).toHaveLength(1);
    expect(validationFailures[0].message).toMatchSnapshot('when parsing full MetaEd and another example - message');
    expect(validationFailures[0].sourceMap).toMatchSnapshot('when parsing full MetaEd and another example - sourceMap');
  });

  it('one more example', () => {
    const inputText = [
      'Begin Namespace extension EXTENSION',
      'Domain Entity Staff\n',
      'documentation "Something"\n',
      'decscriptor EducatorEffectiveness\n',
      'documentation "Indicates the educator effectiveness label as identified by the school district for each educator. The effectiveness labels should be based upon the employees most recent evaluation."\n',
      'is optional\n',
      'comvmon Recognition\n',
      'documentation "Recognition given to a staff member for accomplishments in a co-curricular, or extra-curricular activity."\n',
      'is optional collection\n',
      'End Namespace',
    ].join('\n');

    const validationFailures: Array<ValidationFailure> = [];
    const errorListener = new MetaEdErrorListener(validationFailures);
    buildMetaEd(errorListener, inputText);
    expect(validationFailures).toHaveLength(1);
    expect(validationFailures[0].message).toMatchSnapshot('when parsing full MetaEd one more example - message');
    expect(validationFailures[0].sourceMap).toMatchSnapshot('when parsing full MetaEd one more example - sourceMap');
  });
});
