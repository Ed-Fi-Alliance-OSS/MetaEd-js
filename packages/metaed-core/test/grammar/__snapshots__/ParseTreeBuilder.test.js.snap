// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when parsing full MetaEd and another example: when parsing full MetaEd and another example - message 1`] = `"mismatched input 'comvmon' expecting {'Abstract Entity', 'Association', 'End Namespace', 'Choice', 'Common', 'Descriptor', 'Domain', 'Domain Entity', 'Enumeration', 'Interchange', 'Inline Common', 'Shared Decimal', 'Shared Integer', 'Shared Short', 'Shared String', 'Subdomain', 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'with context'}"`;

exports[`when parsing full MetaEd and another example: when parsing full MetaEd and another example - sourceMap 1`] = `
Object {
  "column": 0,
  "line": 12,
  "tokenText": "comvmon",
}
`;

exports[`when parsing full MetaEd another example: when parsing full MetaEd another example - message 1`] = `"mismatched input 'comvmon' expecting {'Abstract Entity', 'Association', 'End Namespace', 'Choice', 'Common', 'Descriptor', 'Domain', 'Domain Entity', 'Enumeration', 'Interchange', 'Inline Common', 'Shared Decimal', 'Shared Integer', 'Shared Short', 'Shared String', 'Subdomain', 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'with context'}"`;

exports[`when parsing full MetaEd another example: when parsing full MetaEd another example - sourceMap 1`] = `
Object {
  "column": 0,
  "line": 10,
  "tokenText": "comvmon",
}
`;

exports[`when parsing full MetaEd one more example: when parsing full MetaEd one more example - message 1`] = `"mismatched input 'decscriptor' expecting {'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'allow primary key updates'}"`;

exports[`when parsing full MetaEd one more example: when parsing full MetaEd one more example - sourceMap 1`] = `
Object {
  "column": 0,
  "line": 6,
  "tokenText": "decscriptor",
}
`;

exports[`when parsing full MetaEd should not parse with extraneous xyz: when parsing full MetaEd should not parse with extraneous xyz - message 1`] = `"extraneous input 'xyz' expecting {'documentation', METAED_ID}"`;

exports[`when parsing full MetaEd should not parse with extraneous xyz: when parsing full MetaEd should not parse with extraneous xyz - sourceMap 1`] = `
Object {
  "column": 23,
  "line": 5,
  "tokenText": "xyz",
}
`;

exports[`when parsing full MetaEd should not parse with multiple keyword syntax errors: when parsing full MetaEd should not parse with multiple keyword syntax errors - message 1`] = `"mismatched input 'comvmon' expecting {'Abstract Entity', 'Association', 'End Namespace', 'Choice', 'Common', 'Descriptor', 'Domain', 'Domain Entity', 'Enumeration', 'Interchange', 'Inline Common', 'Shared Decimal', 'Shared Integer', 'Shared Short', 'Shared String', 'Subdomain', 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'with context'}"`;

exports[`when parsing full MetaEd should not parse with multiple keyword syntax errors: when parsing full MetaEd should not parse with multiple keyword syntax errors - message 2`] = `"mismatched input 'zcommvzxxon' expecting {'Abstract Entity', 'Association', 'End Namespace', 'Choice', 'Common', 'Descriptor', 'Domain', 'Domain Entity', 'Enumeration', 'Interchange', 'Inline Common', 'Shared Decimal', 'Shared Integer', 'Shared Short', 'Shared String', 'Subdomain', 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'with context'}"`;

exports[`when parsing full MetaEd should not parse with multiple keyword syntax errors: when parsing full MetaEd should not parse with multiple keyword syntax errors - sourceMap 1`] = `
Object {
  "column": 0,
  "line": 10,
  "tokenText": "comvmon",
}
`;

exports[`when parsing full MetaEd should not parse with multiple keyword syntax errors: when parsing full MetaEd should not parse with multiple keyword syntax errors - sourceMap 2`] = `
Object {
  "column": 0,
  "line": 28,
  "tokenText": "zcommvzxxon",
}
`;

exports[`when parsing top level entities and another example: when parsing top level entities and another example - message 1`] = `"mismatched input 'comvmon' expecting {<EOF>, 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'with context'}"`;

exports[`when parsing top level entities and another example: when parsing top level entities and another example - sourceMap 1`] = `
Object {
  "column": 0,
  "line": 11,
  "tokenText": "comvmon",
}
`;

exports[`when parsing top level entities another example: when parsing top level entities another example - message 1`] = `"mismatched input 'comvmon' expecting {<EOF>, 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'with context'}"`;

exports[`when parsing top level entities another example: when parsing top level entities another example - sourceMap 1`] = `
Object {
  "column": 0,
  "line": 9,
  "tokenText": "comvmon",
}
`;

exports[`when parsing top level entities one more example: when parsing top level entities one more example - message 1`] = `"mismatched input 'decscriptor' expecting {'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'allow primary key updates'}"`;

exports[`when parsing top level entities one more example: when parsing top level entities one more example - sourceMap 1`] = `
Object {
  "column": 0,
  "line": 5,
  "tokenText": "decscriptor",
}
`;

exports[`when parsing top level entities should not parse with extraneous xyz: when parsing top level entities should not parse with extraneous xyz - message 1`] = `"extraneous input 'xyz' expecting {'documentation', METAED_ID}"`;

exports[`when parsing top level entities should not parse with extraneous xyz: when parsing top level entities should not parse with extraneous xyz - sourceMap 1`] = `
Object {
  "column": 23,
  "line": 4,
  "tokenText": "xyz",
}
`;

exports[`when parsing top level entities should not parse with multiple keyword syntax errors: when parsing top level entities should not parse with multiple keyword syntax errors - message 1`] = `"mismatched input 'comvmon' expecting {<EOF>, 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'with context'}"`;

exports[`when parsing top level entities should not parse with multiple keyword syntax errors: when parsing top level entities should not parse with multiple keyword syntax errors - message 2`] = `"mismatched input 'zcommvzxxon' expecting {<EOF>, 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'with context'}"`;

exports[`when parsing top level entities should not parse with multiple keyword syntax errors: when parsing top level entities should not parse with multiple keyword syntax errors - sourceMap 1`] = `
Object {
  "column": 0,
  "line": 9,
  "tokenText": "comvmon",
}
`;

exports[`when parsing top level entities should not parse with multiple keyword syntax errors: when parsing top level entities should not parse with multiple keyword syntax errors - sourceMap 2`] = `
Object {
  "column": 0,
  "line": 27,
  "tokenText": "zcommvzxxon",
}
`;
